### ISSUE TITLE
Integration layer does not allow Scoped Middlewares - ASPNetCore

### ISSUE THREAD

**Issue by RamjotSingh** [Microsoft/botbuilder-dotnet#357]
Few issues :-
1) IBot is registered under Transient when it should ideally be Scoped.

2) Another issue is that IMiddlewares can't be registered with Scoped model since Options asks for the Middleware instances Middlewares can't be injected based on scopes (I am referring to ASPNetCore inbuilt DI).
FYI @drub0y

**Comment by RamjotSingh** [comment]
I realized this problem stems from the fact that BotFrameworkAdapter is initialized only once at application start in Configure method. Since post that the DI container is not referred to again nothing registered in the DI container will ever be instantiated again.

**Comment by drub0y** [comment]
Yup. FWIW, I totally want to enable this, but right now the overall framework is not DI aware. I am working on bringing DI related to another set of work, but it remains to be seen if it will be wholeheartedly accepted. Ultimately, once DI is supported from "the top down", you'd be able to configure instancing semantics for middleware.

**Comment by cleemullins** [comment]
@RamjotSingh @drub0y 

By design - the SDK should fully embrace people using DI with the Bot Framework. We're avoiding using DI within the Bot Framework, as it would mean pushing our DI preferences (or lack thereof!) onto other people. As we saw in the V3 SDK with the selection of AutoFaq, that choice can be very polarizing. 

If something we're doing is PREVENTING folks from using DI in standard ways, then we should change that. 

IF there are specific changes we could make to better enable DI, please let us know.

**Comment by ukphillips** [comment]
Just now running into this. I want to use dependency injection for creating my middleware (TranscriptLoggerMiddleware) in this case, where I have created my ITranscriptLogger that I need to pass in, which itself gets an internal storage wrapper class passed in to it's constructor. It feels completely out of place to have to create a chain of 3 news:

``options.Middleware.Add(new TranscriptLoggerMiddleware(new MyTranscriptLogger(new MyStorageWrapper(connectionString))));``

I would think I could just plug a type into the pipeline like standard .net core:

``builder.UseMiddleware<TranscriptLoggerMiddleware>();``

given that I had appropriately registered all the dependent types it would just resolve based on the configured scope. I see this is marked as approved/backlog, so hopefully this is the approach that is being taken.

**Comment by johnataylor** [comment]
Couple of options here:

1) In 4.1 we added a new interface IAdapterIntegration so you can use dependency injection to have the integration pieces use a custom BotAdapter. Now can set up your BotAdapter how you like, it could either be completely custom or a custom initialization of the BotFrameworkAdapter.

1) Also take a look at the new [ASP MVC sample](https://github.com/Microsoft/BotBuilder-Samples/tree/master/samples/csharp_dotnetcore/30.asp-mvc-bot) that demonstrates how you can use Bot Builder runtime inside a regular MVC project.

**Comment by jvanderbiest** [comment]
@johnataylor Is there any example on how to setup a custom botadapter using the IAdapterIntegration and let it work together with depenendency injection in your middleware?

**Comment by hansmbakker** [comment]
@cleemullins @RamjotSingh @drub0y 

> By design - the SDK should fully embrace people using DI with the Bot Framework. We're avoiding using DI within the Bot Framework, as it would mean pushing our DI preferences (or lack thereof!) onto other people. As we saw in the V3 SDK with the selection of AutoFaq, that choice can be very polarizing.

Can't you implement best practices using [Microsoft.Extensions.DependencyInjection.Abstractions](https://www.nuget.org/packages/Microsoft.Extensions.DependencyInjection.Abstractions/)? I believe that then people have a choice to use that in combination with their preferred container.

E.g. using https://github.com/unitycontainer/microsoft-dependency-injection or https://github.com/autofac/Autofac.Extensions.DependencyInjection

**Comment by drub0y** [comment]
@hansmbakker Yeah that's exactly what I started to do starting at the ASP.NET integration layer and, at bare minimum, you can count on dependencies being injected into your `IBot`, but the core framework itself and things like the Dialogs API do not participate or even lend themselves to playing well with DI today. There are actually some things I could do to enable scoped middleware execution and I've even found a way to open up the Dialogs API ([prototype work here](https://github.com/Microsoft/botbuilder-dotnet/tree/drmarsh/dcr/better-dialogs-integration)) to enable support for lifetime scoping and enable full support for DI in a completely backwards compatible way, but I'm finding it to be a bit of an uphill battle as you can see from the other issue you just commented on as well (#1188). Hearing more from people in the .NET community like yourself certainly helps to bolster the case, so please continue to share your thoughts and comments on this topic. 👍

**Comment by hansmbakker** [comment]
@drub0y I implemented a simplified version of DI for Dialogs in https://github.com/Microsoft/AI/pull/1119 based on https://github.com/Microsoft/BotBuilder-Samples/blob/master/samples/csharp_dotnetcore/42.scaleout/Startup.cs
